### ISSUE TITLE
LoadError when deploying to JRuby/Torquebox on Windows

### ISSUE THREAD

**Issue by fpauser** [sparklemotion/nokogiri#1114]
Ruby: JRuby 1.7.11
Rails: 4.0.5
Nokogiri: 1.6.2.1
OS: Windows 2008 Server (64bit)
Torquebox: 1.3.1
Java: 1.7.0_40-b43

Starting the rails-app gives me:

    (LoadError) load error: nokogiri/nokogiri -- java.lang.NoClassDefFoundError: com/sun/org/apache/xpath/internal/VariableStack

Older versions of Nokogiri (tried it with 1.5.11) work.

**Comment by fpauser** [comment]
Any news on that?

**Comment by flavorjones** [comment]
@fpauser Unfortunately, to my knowledge, nobody on nokogiri-core has reproduced this issue. Though we don't need to reproduce it to follow @bbrowning's advice to remove the dependency on the internal classes, it's hard to know we're doing the right thing if we can't reproduce it.

Can anyone provide instructions on how to build an environment on which I can reproduce? Or, for  extra hugs, provide a docker image? If we can get unblocked on repro we'll fix it.

**Comment by flavorjones** [comment]
Should I be able to reproduce this if I start from https://hub.docker.com/_/websphere-liberty/ ?

**Comment by atambo** [comment]
I think you can reproduce it by using the IBM JDK: http://www.ibm.com/developerworks/java/jdk/ instead of the Oracle JDK but you'd have to be using linux.

**Comment by karlhe** [comment]
Or Windows. IBM JDK doesn't support mac environment though.

**Comment by altmind** [comment]
Hit the same issue.

Fresh server, no leftover rubies, gems of custom java libs.
Debian 6.0(oldstable)
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode) - official package from http://ppa.launchpad.net/webupd8team/java/ubuntu
torquebox-dist-3.1.2
Fresh rails app 4.2.4(basically only 1 scaffolded model)

Running torquebox with app deployed gives loooong stacktrace with "org.jruby.exceptions.RaiseException: (LoadError) load error: nokogiri/nokogiri -- java.lang.NoClassDefFoundError: com/sun/org/apache/xpath/internal/jaxp/JAXPExtensionsProvider"

Full torquebox error: https://gist.github.com/altmind/e4452843a16c58ad6d06

Fixing nokogiri to 1.6.1 alleviates the problem

**Comment by ianderse** [comment]
Any updates on this issue? 
Running into this with torquebox 3.1.2 and Redhat enterprise linux 5.11

**Comment by jvshahid** [comment]
I'll discuss with @flavorjones and hopefully we can get something out the doors soon. Sorry for the inconvenience everyone. I'm afraid, this could mean reverting back a few commits that would cause performance degradation.

**Comment by ianderse** [comment]
Thanks for the quick reply @jvshahid, looks like this pull request potentially addresses it? #1212

**Comment by jvshahid** [comment]
I just merged in #1212. Can someone verify that master work and that the problem is fixed.

**Comment by ianderse** [comment]
Trying to test it.
When I attempt to bundle using the master branch with jruby I get:
"Could not find gem 'nokogiri (>= 0) java' in
git://github.com/sparklemotion/nokogiri.git (at master).
Source does not contain any versions of 'nokogiri (>= 0) java'"

**Comment by larskanis** [comment]
@ianderse You must use this procedure:
```sh
git clone https://github.com/sparklemotion/nokogiri
cd nokogiri
bundle
rake gem
gem inst pkg/nokogiri-1.6.7.rc4-java.gem
```

**Comment by ianderse** [comment]
@larskanis @jvshahid using nokogiri-1.6.7.rc4 I get the following error:
"LoadError: no such file to load -- serializer
          require at org/jruby/RubyKernel.java:1071
           (root) at /Users/*/.rvm/gems/jruby-1.7.19@member-portal/gems/nokogiri-1.6.7.rc4-java/lib/nokogiri.rb:22"

**Comment by jvshahid** [comment]
@ianderse can you describe the full environment where you got this error, e.g. output of `java -version`, name of distro you are using, a docker image that we can use to repro the issue.

**Comment by ianderse** [comment]
@jvshahid I can provide basic information right now, I can work on getting more later.

Mac OS X 10.10.5
java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.40-b25, mixed mode)
jruby 1.7.19

I can confirm that I do not get this error when using Nokogiri 1.6.6.4.

The full backtrace:
LoadError: no such file to load -- serializer
          require at org/jruby/RubyKernel.java:1071
           (root) at /Users/ianderse/.rvm/gems/jruby-1.7.19@_/gems/nokogiri-1.6.7.rc4-java/lib/nokogiri.rb:22
          require at org/jruby/RubyKernel.java:1071
           (root) at /Users/ianderse/.rvm/gems/jruby-1.7.19@_l/gems/ruby-saml-0.9.2/lib/onelogin/ruby-saml/saml_message.rb:1
          require at org/jruby/RubyKernel.java:1071
           (root) at /Users/ianderse/.rvm/gems/jruby-1.7.19@_/gems/ruby-saml-0.9.2/lib/onelogin/ruby-saml/saml_message.rb:4

saml_message.rb:4 is require 'nokogiri'
saml_message.rb:1 is require 'cgi'

**Comment by jvshahid** [comment]
I was able to reproduce the error. Thanks @ianderse

**Comment by jvshahid** [comment]
The new jars were missing from the gem, master should be fixed now.

**Comment by ianderse** [comment]
@jvshahid when I build the gem I get the following error:

rake gem
install -c tmp/java/nokogiri/nokogiri.jar lib/nokogiri/nokogiri.jar
cd pkg/nokogiri-1.6.7.rc4-java
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
rake aborted!
Gem::InvalidSpecificationException: ["lib/serializer.jar", "lib/xalan.jar", "lib/xml-apis.jar"] are not files
/Users/ianderse/.rvm/gems/jruby-1.7.19@_/bin/jruby_executable_hooks:15:in `(root)'
Tasks: TOP => gem => pkg/nokogiri-1.6.7.rc4-java.gem

**Comment by jvshahid** [comment]
@ianderse try to run `rake clean`, this should fix the issue.

**Comment by ianderse** [comment]
@jvshahid that successfully allowed me to build the gem on my local environment, in one of my cloud environments I get the following error when trying to build the gem:
"11. ERROR in ext/java/nokogiri/HtmlSaxPushParser.java (at line 181)
	public Thread newThread(Runnable r) {
	              ^^^^^^^^^^^^^^^^^^^^^
The method newThread(Runnable) of type new ThreadFactory(){} must override a superclass method
----------
12. ERROR in ext/java/nokogiri/HtmlSaxPushParser.java (at line 223)
	public HtmlSaxParserContext call() throws Exception {
	                            ^^^^^^^^^^^^^^^^^^^^^^^
The method call() of type HtmlSaxPushParser.ParserTask must override a superclass method
----------
12 problems (2 errors, 10 warnings)rake aborted!
Command failed with status (255): [javac -extdirs "/usr/java/jdk1.7.0_60/jre/...]"

Environment:
JRuby 1.7.19
Redhat enterprise linux 5.11
java version "1.7.0_60"
Java(TM) SE Runtime Environment (build 1.7.0_60-b19)
Java HotSpot(TM) 64-Bit Server VM (build 24.60-b09, mixed mode)

I will try building it locally and testing it that way, but thought you should know about this error as well.

**Comment by ianderse** [comment]
@jvshahid based on my initial tests, I can confirm that after your recent fix I am able to deploy and no longer seeing the original issues in this thread. Thanks for the hard work!

Environment:
JRuby 1.7.19
Redhat enterprise linux 5.11
java version "1.7.0_60"
Java(TM) SE Runtime Environment (build 1.7.0_60-b19)
Java HotSpot(TM) 64-Bit Server VM (build 24.60-b09, mixed mode)

**Comment by jvshahid** [comment]
Great. Should I be concerned about the 2 errors you reported earlier ?

**Comment by ianderse** [comment]
I'm not sure to be honest, I think as long as people are ok with not being able to build the gem themselves in some environments, it's not a big deal. It's also possible that issue was present in previous versions as well as this was the first time I tried building it in that environment.

**Comment by jvshahid** [comment]
I'm downloading the jdk 7u60 from oracle and will try to build the gem using it instead of openjdk. I checked the java docs and everything seems right, not sure why you are seeing this error. Could it be multiple jdk versions on your server ?

Do you mind opening a new issue with the error you got and the details you posted in the previous comment. That way we can keep track of it and I can ping you if I need more details later.

I'll go ahead and close this issue since you think master resolved the issue.

**Comment by jvshahid** [comment]
Fixed by #1212, #1380 and https://github.com/sparklemotion/nokogiri/commit/d811ea4636741e8785eee9244de60c4522d68573
